miniscope_io/models/pipeline.py

```diff
@@ -0,0 +1,250 @@
+
+        .. note::
+
+            The `process` method should not directly be called or passed data,
+            but instead should be called by a containing pipeline class.
+
+        """
+
+
+class ProcessingNode(Node, Generic[T, U]):
+    """
+    An intermediate processing node that transforms some input to output
+    """
+
+    @abstractmethod
+    def process(self, data: T) -> U:
+        """
+        Process some incoming data, yielding a transformed output
+
+        .. note::
+
+            The `process` method should not directly call or be called by
+            output or input nodes, but instead return the output
+            and allow a containing pipeline class to handle dispatching data.
+
+        """
+
+
+class Pipeline(PipelineModel):
+    """
+    A graph of nodes transforming some input source(s) to some output sink(s)
+    """
+
+    nodes: dict[str, Node] = Field(default_factory=dict)
+    """
+    Dictionary mapping all nodes from their ID to the instantiated node.
+    """
+
+    @property
+    def sources(self) -> dict[str, "Source"]:
+        """All :class:`.Source` nodes in the processing graph"""
+        return {k: v for k, v in self.nodes.items() if isinstance(v, Source)}
+
+    @property
+    def processing_nodes(self) -> dict[str, "ProcessingNode"]:
+        """All :class:`.ProcessingNode` s in the processing graph"""
+        return {k: v for k, v in self.nodes.items() if isinstance(v, ProcessingNode)}
+
+    @property
+    def sinks(self) -> dict[str, "Sink"]:
+        """All :class:`.Sink` nodes in the processing graph"""
+        return {k: v for k, v in self.nodes.items() if isinstance(v, Sink)}
+
+    @abstractmethod
+    def process(self) -> None:
+        """
+        Process one step of data from each of the sources,
+        passing intermediate data to any subscribed nodes in a chain.
+
+        The `process` method should not return anything except a to-be-implemented
+        result/status object, as any data intended to be received/processed by
+        downstream objects should be accessed via a :class:`.Sink` .
+        """
+
+    @abstractmethod
+    def start(self) -> None:
+        """
+        Start processing data with the pipeline graph
+        """
+
+    @abstractmethod
+    def stop(self) -> None:
+        """
+        Stop processing data with the pipeline graph
+        """
+
+    @classmethod
+    def from_config(cls, config: PipelineConfig) -> Self:
+        """
+        Instantiate a pipeline model from its configuration
+        """
+        types = Node.node_types()
+
+        nodes = {k: types[v.type_].from_config(v) for k, v in config.nodes.items()}
+        nodes = connect_nodes(nodes)
+        return cls(nodes=nodes)
+
+
+def connect_nodes(nodes: dict[str, Node]) -> dict[str, Node]:
+    """
+    Provide references to instantiated nodes
+    """
+
+    for node in nodes.values():
+        if node.config.inputs and node.inputs is None:
```

Comment:
Will further verification only happen in the node inputs? Or will we handle it somehow in connecting modules?

Reply:
Verification for e.g. the input and output types matching? ya that will be TBD, this is just a stub and i was planning on working on that as part of the first implementation as i move the SDCard classes over.